### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by jbergknoff** [aws/amazon-ecs-agent#362]
Lately I've run into an issue where the ECS agent fails to start a task, stating something like

```
CannotPullECRContainerError: RequestError: send request failed caused by: Post https://ecr.us-east-1.amazonaws.com/: dial tcp: lookup ecr.us-east-1.amazonaws.com on 10.0.0.2:53: dial udp 10.0.0.2:53: too many open files
```

or

```
CannotPullContainerError: dial unix /var/run/docker.sock: too many open files
```

Agent logs contain things like

```
016-04-05T19:49:37Z [ERROR] Error getting libcontainer stats module="stats" err="failed to parse memory.usage_in_bytes - open /cgroup/memory/docker/e71ad63b0bc740067df517562321a224925731bb8eb40c68afda93c11495b15a/memory.usage_in_bytes: too many open files" 
```

```
2016-04-05T19:49:28Z [WARN] Error with docker; stopping container module="TaskEngine" task="..., Status: (NONE->RUNNING) Containers: [... (CREATED->RUNNING),]" container="... (CREATED->RUNNING)" err="dial unix /var/run/docker.sock: too many open files" 
```

I SSHed into the instance (running amzn-ami-2015.09.g-amazon-ecs-optimized (ami-33b48a59), Docker 1.9.1, agent 1.8.1) and found

```
$ ps aux | grep agent
root      3039  0.9  1.6  70160 66772 ?        Ssl  Apr04  12:58 /agent
$ cat /proc/3039/limits
Limit                     Soft Limit           Hard Limit           Units     
...
Max open files            1024                 4096                 files
$ sudo ls /proc/3039/fd | wc -l
1024
```

In this state, the instance is unusable, `docker ps` hangs and no tasks can start. Any ideas what gets the agent into this state, how to avoid it, how to fix it, etc.? Thanks.

**Comment by kiranmeduri** [comment]
Hi  __jbergknoff@__ thanks for reporting this issue. I tried to reproduce this but no luck. There certainly seems like connection leak somewhere. Can you provide __sanitized__ agent logs? Also it will be great to understand the context.

- How many tasks are running on this instance, when this issue shows up?
- How frequently does this happen? 
- Does the instance eventually get into this state or enters into this immediately?

Thanks,
-kiran

**Comment by jbergknoff** [comment]
Unfortunately the instances where this happened have been terminated. I will try to provide you with more details if we encounter it again.

The instance does not begin in this state. Our workload involves many short-lived tasks that are started en masse on a schedule, and the task activity seemed to trigger this state on some (but not all) instances in the cluster.

This only began happening a couple of weeks ago. Two thoughts:

* 3 weeks ago we upgraded from `amzn-ami-2015.09.e-amazon-ecs-optimized` to `amzn-ami-2015.09.g-amazon-ecs-optimized`. A couple of days ago, we upgraded to `amzn-ami-2016.03.a-amazon-ecs-optimized` and I believe we have not seen the issue since then.

* Maybe some of our recently pushed code causes the issue by mismanagement of file or socket IO. If that were the case, is there any chance it could affect the agent's behavior in this way?

**Comment by samuelkarp** [comment]
@jbergknoff This may ultimately be related to https://github.com/aws/amazon-ecs-agent/issues/301.  We've changed the way storage is configured in the new 2016.03.a AMI to address that problem, and it may have helped here too.  Since it has been a few days, are you running into any trouble with the new AMI or has it resolved your issue?

**Comment by jbergknoff** [comment]
I haven't experienced this issue with the new AMI. Thanks, I'll close this issue.

**Comment by pgeez** [comment]
Hello. I'm on AMI `amzn-ami-2016.03.f` and I'm still experiencing this.

```
2016-08-30T09:54:19Z [WARN] Error retrieving stats for container 880a95591a981b70b66f412cc38443728670ce2bc5d4ef598fe6c32aa7b2cd03: dial unix /var/run/docker.sock: socket: too many open files
2016-08-30T09:54:19Z [WARN] Error retrieving stats for container 0602c7d9eecb6c65aa79e5c5531cf27b7a58bb9e534c25468c2036ba88baea0c: dial unix /var/run/docker.sock: socket: too many open files
2016-08-30T09:54:19Z [WARN] Error retrieving stats for container 880a95591a981b70b66f412cc38443728670ce2bc5d4ef598fe6c32aa7b2cd03: dial unix /var/run/docker.sock: socket: too many open files
2016-08-30T09:54:19Z [WARN] Error retrieving stats for container 0602c7d9eecb6c65aa79e5c5531cf27b7a58bb9e534c25468c2036ba88baea0c: dial unix /var/run/docker.sock: socket: too many open files
2016-08-30T09:54:19Z [WARN] Error retrieving stats for container 880a95591a981b70b66f412cc38443728670ce2bc5d4ef598fe6c32aa7b2cd03: dial unix /var/run/docker.sock: socket: too many open files
```

Worse off, because the ECS agent is logging with a sub-second resolution (check the timestamps), agent logs are now 3GB and have consumed all of the space on the disk.

![image](https://cloud.githubusercontent.com/assets/1369475/18104018/2b9c2a04-6ec7-11e6-8373-095ba313efb6.png)


Please reopen this issue as it does not appear to be fixed.

**Comment by jbergknoff** [comment]
@pgeez this fix appeared in the 2016.03.h AMI.

**Comment by samuelkarp** [comment]
@pgeez Please either launch the new 2016.03.h AMI or upgrade your existing instance(s) to agent version 1.12.1.
